srcbpatch/streamreplacer.cpp

```diff
@@ -0,0 +1,544 @@
+
+public:
+    void SetNextReplacer(StreamReplacer* const pNext) override
+    {
+        pNext_ = pNext;
+    }
+};
+
+
+//--------------------------------------------------
+struct ReplacerPairHolder
+{
+    ReplacerPairHolder(unique_ptr<AbstractBinaryLexeme>& src,  // what to replace
+        unique_ptr<AbstractBinaryLexeme>& trg)                 // to what
+        : src_(src)
+        , trg_(trg)
+    {}
+
+    unique_ptr<AbstractBinaryLexeme>& src_; // what to replace
+    unique_ptr<AbstractBinaryLexeme>& trg_; // with what
+};
+
+
+//--------------------------------------------------
+class UsualReplacer final : public ReplacerWithNext
+{
+public:
+    UsualReplacer(unique_ptr<AbstractBinaryLexeme>& src,  // what to replace
+        unique_ptr<AbstractBinaryLexeme>& trg)  // with what
+        : src_(src->access())
+        , trg_(trg->access())
+    {
+        buffer_.resize(src_.size() + 1); // reserve place for usage (+1 needed for inner logic but never being accessed)
+    }
+
+    void DoReplacements(const char toProcess, const bool aEod) const override;
+
+protected:
+    const span<const char>& src_; // what to replace
+    const span<const char>& trg_; // with what
+
+    mutable size_t cachedAmount_ = 0; // we cached this amount of data
+
+    // this is used to hold temporary data while the logic is 
+    // looking for the new beginning of the cached value
+    mutable vector<char> buffer_;
+};
+
+
+void UsualReplacer::DoReplacements(const char toProcess, const bool aEod) const
+{
+    if (nullptr == pNext_)
+    {
+        throw logic_error("Replacement chain has been broken. Communicate with maintainer");
+    }
+
+    // no more data
+    // just send cached amount
+    if (aEod)
+    {
+        for (size_t i = 0; i < cachedAmount_; ++i)
+        {
+            pNext_->DoReplacements(src_[i], false);
+        }
+        cachedAmount_ = 0;
+        pNext_->DoReplacements(toProcess, true);
+        return;
+    }
+
+    if (src_[cachedAmount_] == toProcess) // check for match
+    {
+        if (++cachedAmount_ >= src_.size())
+        {// send target - do replacement
+            for (size_t q = 0; q < trg_.size(); ++q) { pNext_->DoReplacements(trg_[q], false); }
+            cachedAmount_ = 0;
+        }
+        return;
+    }
+
+    // here:   toProcess is not our char
+    //         lets check for fast track (255/256 probability)
+    if (0 == cachedAmount_)
+    {
+        pNext_->DoReplacements(toProcess, false);
+        return;
+    }
+
+    // here:   We have some cached data
+    //         at least 1 char need to be send further
+    //         remaining cached data including toProcess need to be reprocessed for match
+
+    memcpy(buffer_.data(), src_.data(), cachedAmount_);
+    buffer_[cachedAmount_++]= toProcess;
+    size_t i = 0;
+    do
+    {
+        pNext_->DoReplacements(buffer_[i++], false); // send 1 byte after another
+    } while (0 != memcmp(src_.data(), &buffer_[i], --cachedAmount_));
```

Comment:
Some magic going here. Let's have a look for an example. Let's say that 
src = "abcdefghijk"
cachedAmount_ = 6
toProcess = 's'

so our buffer_ = "abcdef" + toProcess = "abcdefs"
cachedAmount_ = 7

now we are inside the loop

after first do scope : cachedAmount_ = 6, and in memcmp we compare "abcdef" "bcdefs" 
after second do scope : cachedAmount_ = 5, and in memcmp we compare "abcde" "cdefs" 
and so on that's doesn't make sense because every time they are not equal

Or ..... pNext_->DoReplacements(buffer_[i++], false); 
do modify out src_ ????? Then it's non transparent logic.

Reply:
src_ is untouched.
buffer
let's assume:  src_ is "aaaak";  cachedAmount_ == 4
we got 'a' again. so the sequence now "**a**aaaa"
 as a result we still need to have  cachedAmount_ == 4, but first 'a' need to be sent

Reply:
Yeah, I see your point, I just fell to thinking about optimization here. If cachedAmount_ is large and we know that there are no matches... We will still do memcpy for no reason. 

I can't predict it, may be that case I have described above is too rare and you don't have to care about it.